internal/controller/pkg/transaction/installer.go

```diff
@@ -0,0 +1,712 @@
+		return errors.Wrap(err, "cannot list packages")
+	}
+
+	// Use existing package name if found, otherwise keep generated name
+	if name := FindExistingPackage(pkgs, xp); name != "" {
+		pkg.SetName(name)
+	}
+
+	_, err = ctrl.CreateOrUpdate(ctx, i.kube, pkg, func() error {
+		pkg.SetSource(xpkg.BuildReference(xp.Source, xp.Version))
+
+		// For new packages, generation is 0 on the client side but will be 1
+		// after the API server creates them. For existing packages, use their
+		// current generation.
+		generation := pkg.GetGeneration()
+		if generation == 0 {
+			generation = 1
+		}
+
+		meta.AddLabels(pkg, map[string]string{
+			v1alpha1.LabelTransactionName:       tx.GetName(),
+			v1alpha1.LabelTransactionGeneration: strconv.FormatInt(generation, 10),
+		})
+
+		return nil
+	})
+
+	return errors.Wrap(err, "cannot create or update package")
+}
+
+// RevisionCreator creates or updates PackageRevision resources.
+type RevisionCreator struct {
+	kube client.Client
+}
+
+// NewRevisionCreator returns a new RevisionCreator.
+func NewRevisionCreator(kube client.Client) *RevisionCreator {
+	return &RevisionCreator{kube: kube}
+}
+
+// Install installs a PackageRevision by creating or updating it,
+// deactivating other revisions, and garbage collecting old revisions.
+//
+// The package manager maintains at most one active revision per package at any
+// time. When a new revision is installed, all other revisions are deactivated.
+// This ensures a clean transition between package versions.
+func (i *RevisionCreator) Install(ctx context.Context, tx *v1alpha1.Transaction, xp *xpkg.Package) error {
+	pkg, rev, err := NewPackageAndRevision(xp)
+	if err != nil {
+		return err
+	}
+
+	// List packages to find pkgs one with matching source repository
+	pkgs, revs, err := NewPackageAndRevisionList(xp)
+	if err != nil {
+		return err
+	}
+	if err := i.kube.List(ctx, pkgs); err != nil {
+		return errors.Wrap(err, "cannot list packages")
+	}
+
+	// Use existing package name if found, otherwise keep generated name
+	if name := FindExistingPackage(pkgs, xp); name != "" {
+		pkg.SetName(name)
+		rev.SetName(xpkg.FriendlyID(name, xp.DigestHex()))
+	}
+
+	if err := i.kube.Get(ctx, types.NamespacedName{Name: pkg.GetName()}, pkg); err != nil {
+		return errors.Wrap(err, "cannot get package")
+	}
+
+	if err := i.kube.List(ctx, revs, client.MatchingLabels{v1.LabelParentPackage: pkg.GetName()}); err != nil {
+		return errors.Wrap(err, "cannot list package revisions")
+	}
+
+	// Find the highest revision number to ensure our new revision gets a
+	// higher number. Revision numbers are monotonically increasing.
+	var maxRevision int64
+	for _, r := range revs.GetRevisions() {
+		if r.GetRevision() > maxRevision {
+			maxRevision = r.GetRevision()
+		}
+	}
+
+	// Deactivate all other revisions before activating the new one. This
+	// ensures only one revision is active at a time, regardless of the
+	// package's activation policy.
```

Comment:
Note that this leaves things in a weird state (no revision marked active) when using the manual activation policy. If we move forward with this, I think it would be worth merging/integrating #6446 first to clean up activation.